Enhancements for the "Compare against ..."

deathaxe opened the pull request:
I'd like to introduce the following changes to GitGutter. I packed several things into this PR as they all belong together, modify the same parts of code and would otherwise make too much work to merge correctly again as they base on each other.

### 1 Infrastructure
The first commits are some kind of infrastructure changes to ease handling of git output and reduce some cross references of git_gutter modules. They mainly intend to reduce required lines of code. They are the base of the following functional changes and additions.

### 2 Show branch name in compared against status message

I was always confuesed about the Comparing against: ... status messages. It only shows HEAD or a certain commit hash instead of the name of a branch selected as compare target. 

#### Issue

Comparing against branch doesn't actually tell GitGutter to compare against the HEAD of the selected branch but only against the currently latest commit. This has two side effects. The branch name is not displayed in the status bar and if the branch forwards, GitGutter won't see that.

#### Solution

Save the whole `refs/heads/...` or `refs/tags/...` strings instead of commit hashes, if _Compare against branch_ or _Compare against tag_ was called.

### 3 Make the status message somewhat smarter

The status message should show for all files within a git repository no matter if the gutter icons show or not. So user can see, if the file is contained in a git repo and which state (ignored/untracked) it has. Furthermore the statistics message itself should always show the 'Regions' or 'Lines' extension after the first visible _inserted/deleted/modified_ item.

### 4 Compare against specific file commit (Issue #281) 

Add a new function to select all commits sorted by author date, which introduced changes to the file displayed in the active view.

rchl commented on git_gutter_handler.py:
```diff
@@ -162,7 +162,7 @@ def write_file(contents):
             try:
                 self.update_git_time()
-                return GitGutterHandler.run_command(args).then(write_file)
+                return GitGutterHandler.run_command(args, False).then(write_file)
```
Nit: Would be clearer if you would write `decode=False`.

rchl commented on git_gutter_handler.py:
```diff
@@ -229,7 +229,7 @@ def run_diff(_unused):
             ]
             args = list(filter(None, args))  # Remove empty args
-            return GitGutterHandler.run_command(args).then(decode_diff)
+            return GitGutterHandler.run_command(args, False).then(decode_diff)
```
Nit: Would be clearer if you would write `decode=False`.

deathaxe commented:
Unfortunatelly there are some overlaps as this PR is somewhat older. But you are right. Let's check the other one first.

deathaxe commented:
Rabased the work on new master state and already worked in the handful of tips from @rchl.

r-stein commented:
The statusbar is a little bit to verbose and a little inconsistent:
```
Comparing against: HEAD, File is modified, Deleted : 1 region, Inserted : 3 lines, Modified: 5
```
We may make the "Comparing against" and "File is modified/added/commited" optional and maybe add the branch and shorted the rest with symbols like:
```
1-, 3+, 5≠
```

deathaxe commented:
This is exactly the feedback I need! 
1. I aggree with you that all information make it quite long. I already played around with hiding the 'File is modified' and show 'added/staged/commited' states only. This is what it behaves on my local development branch right now.
2. To show 'added/staged' states and efficiently query 'ignored/untracked' files and branch name with a single git call, I created a patch to call 
```python
        args = [
            settings.git_binary_path,
            '--icase-pathspecs',
            'status',
            '--porcelain', '-b', '-s', '-u', '--ignored',
            self._git_path
        ]
```
By now, we need 3 git calls to query branch, ignored state and untracked.
This works pretty well right now, but I am not sattisfied with my handling of the git status yet. This needs some further fine tuning and cleanup. But it is work in progress and is as many things based on my fix for Issue #340  ;-)

For now I may shorten the status message, only. But I need to check, whether ST2 can handle the `≠`

deathaxe commented:
I remember a dictionary like setting was discussed to enable/disable each element separately was discussed somewhere. May I create such a setting? I did not do so right now, because of compatibility concerns.

r-stein commented:
My suggestion would be to create a format string, e.g.:

``` js
"status_bar_text": "Branch: {branch}, Comparing against: {compare_target}, File is {file_state} Deleted: {deleted} region, Inserted : {inserted} lines, Modified: {modified}"
```

Why cou can easily change to 

``` js
"status_bar_text": "On {branch}, {deleted}-, {inserted}+, {modified}≠"
```
I am aware that it currently works with several status, but it does not seem to hard to change that.

With a little documentation this seems to be the easiest and most flexible for the users and for us.
We just run

``` py
current_branch = get_current_branch()
status_bar_text.format(branch=current_branch, deleted=deleted, ...)
```

deathaxe commented:
Sounds great. Will do so.

deathaxe commented:
Using one format string only will make it impossible to hide certain parts of the status message. The statistics would always show as `0+ 0- 0≠` for a commited file.

So we might split the format string in several parts to avoid such glitches.

How about the following settings and their implementation?

```js
  // Format of the status message.
  // Available variables are:
  //   {repo}     repository name / folder name containing the .git directory
  //   {branch}   checked out branch
  //   {compare}  commit/branch/HEAD the file is compared to
  //   {state}    One of added/modified/staged/untracked/ignored
  "status_bar_text_state": "On {branch}, Comparing against {compare}, File {state}",

  // Modification statistics
  //   Available variables are:
  //   {deleted}  number of deleted regions
  //   {inserted} number of inserted lines
  //   {modified} number of modified lines
  "status_bar_text_stats": "{deleted}- {inserted}+ {modified}≠",
```

r-stein commented:
I didn't though about that. My idea would be to support the jinja2 template engine instead. We have it as a ST3 dependency anyway for the popup.

Setting:
``` js
"status_bar_text": "On {{branch}}{% if deleted != 0 %}, {{deleted}}-, {% endif %}{% if inserted != 0 %}, {{inserted}}+{% endif %}{% if modified != 0 %}, {{modified}}≠{% endif %}"
```

Code:

``` py
import jinja2
status_bar_text = "On {{branch}}{% if deleted != 0 %}, {{deleted}}-, {% endif %}{% if inserted != 0 %}, {{inserted}}+{% endif %}{% if modified != 0 %}, {{modified}}≠{% endif %}"
tmpl = jinja2.environment.Template(status_bar_text)
tmpl.render(branch="master", deleted=0, inserted=5, modified=12)
```

This gives a less readable setting, but think it is still not to complicated. Especially since the average user knows git and therefore is most likely a programmer himself.

deathaxe commented:
This is dark magic. Even works for ST2. Just crazy. The pitfall is the very long and maybe complicated template string for some users. Syntax is easy to understand for me, but is it for more genaral users, too? I have no idea how common this template language is.

r-stein commented:
Me must also adept the [dependencies](https://github.com/jisaacks/GitGutter/blob/master/dependencies.json) to require `python-jinja` for ST2.
Not sure about the average user, but we could add a few example setting strings above.

deathaxe commented:
Examples are a good idea. Most people will live with default value anyway, I think.

I would change the setting to an array, so the entries can be split into single lines for better readability.

```js
  "status_bar_text": [
    "On {{branch}}",
    "{% if compare != 'HEAD' %}, Comparing against {{compare}}{% endif %}",
    ", File is {{state}}",
    "{% if deleted != 0 %}, {{deleted}}-{% endif %}",
    "{% if inserted != 0 %}, {{inserted}}+{% endif %}",
    "{% if modified != 0 %}, {{modified}}≠{% endif %}"
  ],
```

The array is joint to a single string and passed to jinja2 Template.

deathaxe commented:
I tend to use `show_status_bar_text:true` instead of `show_status` as it sounds a bit more accurate and may not interfere with the old string setting. What do you mean?

deathaxe commented:
I rebased this branch on the current master and squashed away the old Enhancement: Status bar ... commit to keep only the commits intended to go to master. Please do a final check, if everything is still fine. If so, I will rebase the master on it.
